tweet_recommendations/grid_search.py

```diff
@@ -0,0 +1,112 @@
+
+
+METRICS = {'recall': get_recall_at_k,
+           'mAP': get_average_precision_at_k,
+           'ndcg': get_rank_dcg_at_k}
+
+
+EMBEDDING_NAMES = ['skipgram', 'fasttext']
+POPULARITY_MEASURES = ['pagerank', 'mean_retweets']
+RATIOS = np.linspace(0, 1, 11)
+Ks = [5, 10, 15, 20, 30, 40, 50, 75, 100]
+HASHTAG_OCCURENCE_COUNT_LIMIT = [0, 1, 5, 10, 20]
+
+
+def create_datasets_pool():
+    full_dataset = get_full_dataset()
+    train_df, pop_train_df, unpop_train_df = get_train_dataset(full_dataset)
+    val_df, pop_val_df, unpop_val_df = get_val_dataset(full_dataset)
+    test_df, pop_test_df, unpop_test_df = get_test_dataset(full_dataset)
+
+    return {'train': train_df,
+            'train_popular': pop_train_df,
+            'train_unpopular': unpop_train_df,
+            'val': val_df,
+            'val_popular': pop_val_df,
+            'val_unpopular': unpop_val_df,
+            'test': test_df,
+            'test_popular': pop_test_df,
+            'test_unpopular': unpop_test_df}
+
+
+def build_hyperparameters_dataframe(dataset_names, embeddings, popularities, ratios, ks, count_limits, ):
+    return pd.DataFrame(list(product(dataset_names, embeddings, popularities, ratios, ks, count_limits)),
+                        columns=['dataset', 'embedding', 'popularity', 'ratio', 'K', 'occurence_limit'])
+
+
+def get_metrics_for_recommendation(expected, recommended, metrics_functions, k):
+    return {metric: metrics_functions[metric](expected, recommended, k)
+            for metric in metrics_functions}
+
+
+def experiment(tweets_df, hashtags_df, hyperparameters, metrics_to_calculate=METRICS, backup_dir=None):
+    config = {'popularity_to_similarity_ratio': hyperparameters['ratio'],
+              'embedding_name': hyperparameters['embedding'],
+              'popularity_measure': hyperparameters['popularity'],
+              'K': hyperparameters['K']}
+
+    embedding = hyperparameters['embedding']
+    hashtags = filter_hashtags_by_occurences_limit(hashtags_df, hyperparameters['occurence_limit'])
+    df = compute_expected_hashtags_with_respect_to_hashtags_df(tweets_df, hashtags)
+
+    precomputed_similarities = embedding_similarity(np.asarray(df[embedding].values.tolist()),
+                                                    np.asarray(hashtags[embedding].values.tolist()))
+    df['similarities'] = [precomputed_similarities[i] for i in range(len(precomputed_similarities))]
+
+    df['recommendations'] = df['similarities'].apply(
+        lambda sim: recommend_with_computed_similarities(sim, hashtags, config))
+
+    metrics = df.apply(lambda row: get_metrics_for_recommendation(row['expected_hashtags'], row['recommendations'],
+                                                                  metrics_to_calculate, config['K']),
+                       axis=1)
+    result = pd.DataFrame(list(metrics)).mean()
+
+    del df, hashtags, precomputed_similarities

```

Comment:
Prawdopodobnie to nie wpływa w żaden sposób na program, zwłaszcza, że za chwilę następuje powrót z metody.